## NFL Play Predictor!
#### Predicting whether an NFL team will run or pass depending on game context: down, distance, time, and score.

### Part 1: Importing + Selecting Data

In [2]:
%pip install nflreadpy
import nflreadpy as nfl

Note: you may need to restart the kernel to use updated packages.


In [3]:
# now we can snag the play by play data
pbp = nfl.load_pbp()
pbp.head()

play_id,game_id,old_game_id,home_team,away_team,season_type,week,posteam,posteam_type,defteam,side_of_field,yardline_100,game_date,quarter_seconds_remaining,half_seconds_remaining,game_seconds_remaining,game_half,quarter_end,drive,sp,qtr,down,goal_to_go,time,yrdln,ydstogo,ydsnet,desc,play_type,yards_gained,shotgun,no_huddle,qb_dropback,qb_kneel,qb_spike,qb_scramble,pass_length,…,home_coach,away_coach,stadium_id,game_stadium,aborted_play,success,passer,passer_jersey_number,rusher,rusher_jersey_number,receiver,receiver_jersey_number,pass,rush,first_down,special,play,passer_id,rusher_id,receiver_id,name,jersey_number,id,fantasy_player_name,fantasy_player_id,fantasy,fantasy_id,out_of_bounds,home_opening_kickoff,qb_epa,xyac_epa,xyac_mean_yardage,xyac_median_yardage,xyac_success,xyac_fd,xpass,pass_oe
f64,str,str,str,str,str,i32,str,str,str,str,f64,str,f64,f64,f64,str,f64,f64,f64,f64,f64,i32,str,str,f64,f64,str,str,f64,f64,f64,f64,f64,f64,f64,str,…,str,str,str,str,f64,f64,str,i32,str,i32,str,i32,f64,f64,f64,f64,f64,str,str,str,str,i32,str,str,str,str,str,f64,f64,f64,f64,f64,i32,f64,f64,f64,f64
1.0,"""2025_01_ARI_NO""","""2025090705""","""NO""","""ARI""","""REG""",1,null,null,null,null,null,"""2025-09-07""",900.0,1800.0,3600.0,"""Half1""",0.0,null,0.0,1.0,null,0,"""15:00""","""NO 35""",0.0,null,"""GAME""",null,null,0.0,0.0,null,0.0,0.0,0.0,null,…,"""Kellen Moore""","""Jonathan Gannon""","""NOR00""","""Mercedes-Benz Superdome""",0.0,0.0,null,null,null,null,null,null,0.0,0.0,null,0.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,0.0,-0.0,null,null,null,null,null,null,null
40.0,"""2025_01_ARI_NO""","""2025090705""","""NO""","""ARI""","""REG""",1,"""ARI""","""away""","""NO""","""NO""",35.0,"""2025-09-07""",900.0,1800.0,3600.0,"""Half1""",0.0,1.0,0.0,1.0,null,0,"""15:00""","""NO 35""",0.0,2.0,"""19-B.Grupe kicks 65 yards from…","""kickoff""",0.0,0.0,0.0,0.0,0.0,0.0,0.0,null,…,"""Kellen Moore""","""Jonathan Gannon""","""NOR00""","""Mercedes-Benz Superdome""",0.0,0.0,null,null,null,null,null,null,0.0,0.0,0.0,1.0,0.0,null,null,null,null,null,null,null,null,null,null,0.0,0.0,-0.3527,null,null,null,null,null,null,null
63.0,"""2025_01_ARI_NO""","""2025090705""","""NO""","""ARI""","""REG""",1,"""ARI""","""away""","""NO""","""ARI""",78.0,"""2025-09-07""",896.0,1796.0,3596.0,"""Half1""",0.0,1.0,0.0,1.0,1.0,0,"""14:56""","""ARI 22""",10.0,2.0,"""(14:56) 6-J.Conner right tackl…","""run""",3.0,0.0,0.0,0.0,0.0,0.0,0.0,null,…,"""Kellen Moore""","""Jonathan Gannon""","""NOR00""","""Mercedes-Benz Superdome""",0.0,0.0,null,null,"""J.Conner""",6,null,null,0.0,1.0,0.0,0.0,1.0,null,"""00-0033553""",null,"""J.Conner""",6,"""00-0033553""","""J.Conner""","""00-0033553""","""J.Conner""","""00-0033553""",0.0,0.0,-0.190052,null,null,null,null,null,0.511128,-51.112807
85.0,"""2025_01_ARI_NO""","""2025090705""","""NO""","""ARI""","""REG""",1,"""ARI""","""away""","""NO""","""ARI""",75.0,"""2025-09-07""",858.0,1758.0,3558.0,"""Half1""",0.0,1.0,0.0,1.0,2.0,0,"""14:18""","""ARI 25""",7.0,2.0,"""(14:18) (Shotgun) 1-K.Murray p…","""pass""",11.0,1.0,0.0,1.0,0.0,0.0,0.0,"""short""",…,"""Kellen Moore""","""Jonathan Gannon""","""NOR00""","""Mercedes-Benz Superdome""",0.0,1.0,"""K.Murray""",1,null,null,"""T.McBride""",85,1.0,0.0,1.0,0.0,1.0,"""00-0035228""",null,"""00-0037744""","""K.Murray""",1,"""00-0035228""","""T.McBride""","""00-0037744""","""T.McBride""","""00-0037744""",1.0,0.0,1.31734,0.939998,4.750889,3,0.666726,0.43911,0.66894,33.105969
115.0,"""2025_01_ARI_NO""","""2025090705""","""NO""","""ARI""","""REG""",1,"""ARI""","""away""","""NO""","""ARI""",64.0,"""2025-09-07""",820.0,1720.0,3520.0,"""Half1""",0.0,1.0,0.0,1.0,1.0,0,"""13:40""","""ARI 36""",10.0,2.0,"""(13:40) 1-K.Murray sacked at A…","""pass""",-11.0,0.0,0.0,1.0,0.0,0.0,0.0,null,…,"""Kellen Moore""","""Jonathan Gannon""","""NOR00""","""Mercedes-Benz Superdome""",0.0,0.0,"""K.Murray""",1,null,null,null,null,1.0,0.0,0.0,0.0,1.0,"""00-0035228""",null,null,"""K.Murray""",1,"""00-0035228""",null,null,n

In [4]:
# looking good, let's check the list since it'll be easier to analyze
pbp.columns

['play_id',
 'game_id',
 'old_game_id',
 'home_team',
 'away_team',
 'season_type',
 'week',
 'posteam',
 'posteam_type',
 'defteam',
 'side_of_field',
 'yardline_100',
 'game_date',
 'quarter_seconds_remaining',
 'half_seconds_remaining',
 'game_seconds_remaining',
 'game_half',
 'quarter_end',
 'drive',
 'sp',
 'qtr',
 'down',
 'goal_to_go',
 'time',
 'yrdln',
 'ydstogo',
 'ydsnet',
 'desc',
 'play_type',
 'yards_gained',
 'shotgun',
 'no_huddle',
 'qb_dropback',
 'qb_kneel',
 'qb_spike',
 'qb_scramble',
 'pass_length',
 'pass_location',
 'air_yards',
 'yards_after_catch',
 'run_location',
 'run_gap',
 'field_goal_result',
 'kick_distance',
 'extra_point_result',
 'two_point_conv_result',
 'home_timeouts_remaining',
 'away_timeouts_remaining',
 'timeout',
 'timeout_team',
 'td_team',
 'td_player_name',
 'td_player_id',
 'posteam_timeouts_remaining',
 'defteam_timeouts_remaining',
 'total_home_score',
 'total_away_score',
 'posteam_score',
 'defteam_score',
 'score_differential',
 'po

The list is massive; that's great! We have a lot of features to select from. 

I also see play_type, which will be our target variable.

In [ ]:
initial_features = [
    # core situational features
    'down',
    'ydstogo',
    'yardline_100',
    'goal_to_go',
    'score_differential',
    'qtr',

    # game clock
    'game_seconds_remaining',
    'half_seconds_remaining',
    'game_half',
    'quarter_seconds_remaining',

    # number of timeouts
    'home_timeouts_remaining',
    'away_timeouts_remaining',

    # pre-snap formation
    'shotgun',
    'no_huddle',

    # which team has the ball
    'posteam',
]

target = 'play_type' # this will be filtered to just run and pass

target_pbp = pbp[initial_features]

target_pbp.head()

down,ydstogo,yardline_100,goal_to_go,score_differential,qtr,game_seconds_remaining,half_seconds_remaining,game_half,quarter_seconds_remaining,home_timeouts_remaining,away_timeouts_remaining,shotgun,no_huddle,posteam
f64,f64,f64,i32,f64,f64,f64,f64,str,f64,f64,f64,f64,f64,str
null,0.0,null,0,null,1.0,3600.0,1800.0,"""Half1""",900.0,3.0,3.0,0.0,0.0,null
null,0.0,35.0,0,0.0,1.0,3600.0,1800.0,"""Half1""",900.0,3.0,3.0,0.0,0.0,"""ARI"""
1.0,10.0,78.0,0,0.0,1.0,3596.0,1796.0,"""Half1""",896.0,3.0,3.0,0.0,0.0,"""ARI"""
2.0,7.0,75.0,0,0.0,1.0,3558.0,1758.0,"""Half1""",858.0,3.0,3.0,1.0,0.0,"""ARI"""
1.0,10.0,64.0,0,0.0,1.0,3520.0,1720.0,"""Half1""",820.0,3.0,3.0,0.0,0.0,"""ARI"""


In [6]:
target_pbp.columns

['down',
 'ydstogo',
 'yardline_100',
 'goal_to_go',
 'score_differential',
 'qtr',
 'game_seconds_remaining',
 'half_seconds_remaining',
 'game_half',
 'quarter_seconds_remaining',
 'home_timeouts_remaining',
 'away_timeouts_remaining',
 'shotgun',
 'no_huddle',
 'posteam']

After taking another look at our columns, they look good. 
- Shotgun could be a big differentiator, since typically if the pre-snap formation is in shotgun, it will be a pass, and vice versa (if the play is under-center it is typically a run). 
- Down, score_differential, ydstogo, yardline_100, will also likely be top features.

I'm on the fence about keeping 'half_seconds_remaining' since the model will be able to infer that context from "game_half' and 'game_seconds_remaining'. I do feel that it could have relevance, though. 
- For example, if a team gets a drive going before halftime with low time on the clock, they will be passing on most of the plays to try and get some last-second points in. 
- But there is also the scenario where the team kneels the ball and just takes it to the locker room. 

For now, I'll stick with leaving it in as a feature; if it doesn't provide much then I'll remove it later.

Another thing is that 'posteam' and 'home_timeouts_remaining' + 'away_timeouts_remaining' are not synced up. Posteam tells us who has the ball, and the two timeout metrics tell us how many timeouts each team has. But we need to sync those together so that we know how many timeouts the 'posteam' has.

Thankfully, we can use 'posteam_type' to do that linking, since it tells us if the posteam is the home team or the away team.

### Part 2: Data Cleaning

In [7]:
# first, we'll tackle the target variable
print(pbp['play_type'].unique())

shape: (10,)
Series: 'play_type' [str]
[
	"kickoff"
	"punt"
	"pass"
	"run"
	null
	"extra_point"
	"no_play"
	"qb_spike"
	"field_goal"
	"qb_kneel"
]


Play type has 10 different unique column names (9 if you exclude the 'null').  We just need 'run' and 'pass', so let's isolate those.

In [8]:
print(type(pbp))
print(pbp.head())

<class 'polars.dataframe.frame.DataFrame'>
shape: (5, 372)
┌─────────┬────────────┬────────────┬───────────┬───┬────────────┬─────────┬──────────┬────────────┐
│ play_id ┆ game_id    ┆ old_game_i ┆ home_team ┆ … ┆ xyac_succe ┆ xyac_fd ┆ xpass    ┆ pass_oe    │
│ ---     ┆ ---        ┆ d          ┆ ---       ┆   ┆ ss         ┆ ---     ┆ ---      ┆ ---        │
│ f64     ┆ str        ┆ ---        ┆ str       ┆   ┆ ---        ┆ f64     ┆ f64      ┆ f64        │
│         ┆            ┆ str        ┆           ┆   ┆ f64        ┆         ┆          ┆            │
╞═════════╪════════════╪════════════╪═══════════╪═══╪════════════╪═════════╪══════════╪════════════╡
│ 1.0     ┆ 2025_01_AR ┆ 2025090705 ┆ NO        ┆ … ┆ null       ┆ null    ┆ null     ┆ null       │
│         ┆ I_NO       ┆            ┆           ┆   ┆            ┆         ┆          ┆            │
│ 40.0    ┆ 2025_01_AR ┆ 2025090705 ┆ NO        ┆ … ┆ null       ┆ null    ┆ null     ┆ null       │
│         ┆ I_NO       ┆        

I was having issues with syntax, so I checked pbp -- turns out it is a Polars Dataframe, and not a Pandas Dataframe. Let's change that.

In [9]:
pbp = pbp.to_pandas()

In [10]:
pass_run = pbp[pbp['play_type'].isin(['pass', 'run',])]
print(pass_run['play_type'].unique())
print(pass_run.shape)

['run' 'pass']
(34628, 372)


Now let's check for null values.

In [11]:
print(pass_run[initial_features].isnull().sum())

down                         125
ydstogo                        0
yardline_100                   0
goal_to_go                     0
score_differential             0
qtr                            0
game_seconds_remaining         0
half_seconds_remaining         0
game_half                      0
quarter_seconds_remaining      0
home_timeouts_remaining        0
away_timeouts_remaining        0
shotgun                        0
no_huddle                      0
posteam                        0
dtype: int64


They are exclusively under 'down', so let's look into those.

In [12]:
print(pass_run[pass_run['down'].isnull()][['down', 'play_type', 'desc']].head(5))

      down play_type                                               desc
283    NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
311    NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
350    NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
1319   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 15-P.Mahomes pas...
1670   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 1-T.Tagovailoa p...


It seems that they might all be two-point conversion attempts. If that's the case we can drop them, because those do not have a 'down', hence why they return as null.

In [13]:
print(pass_run[pass_run['down'].isnull()][['down', 'play_type', 'desc']])

       down play_type                                               desc
283     NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
311     NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
350     NaN      pass  TWO-POINT CONVERSION ATTEMPT. 17-J.Allen pass ...
1319    NaN      pass  TWO-POINT CONVERSION ATTEMPT. 15-P.Mahomes pas...
1670    NaN      pass  TWO-POINT CONVERSION ATTEMPT. 1-T.Tagovailoa p...
...     ...       ...                                                ...
45931   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 11-M.Trubisky pa...
45957   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 11-M.Trubisky pa...
45969   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 4-B.Cook pass to...
46771   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 18-C.Williams pa...
48752   NaN      pass  TWO-POINT CONVERSION ATTEMPT. 10-D.Maye pass t...

[125 rows x 3 columns]


Yup, they are all two-point attempts. That means we can drop them cleanly and move on.

In [14]:
pass_run = pass_run[pass_run['down'].notna()]
print(pass_run.shape)

(34503, 372)


Ok now to make sure that every column is the correct data type.

In [15]:
print(pass_run[initial_features].dtypes)

down                         float64
ydstogo                      float64
yardline_100                 float64
goal_to_go                     int32
score_differential           float64
qtr                          float64
game_seconds_remaining       float64
half_seconds_remaining       float64
game_half                     object
quarter_seconds_remaining    float64
home_timeouts_remaining      float64
away_timeouts_remaining      float64
shotgun                      float64
no_huddle                    float64
posteam                       object
dtype: object


Almost every feature is a float, but they actaully don't have to be, since all of their values are whole numbers. Unlike basketball, the game clock does not use decimals. Similarly, there is no way to score a half point in football -- only whole numbers. So converting all of these to int is the move.

The data types are likely floats because pandas automatically converts integer columns to flaot64 whenever they contain NaN values, since integers cannot contain NaN in pandas. We have since dropped the nulls, though, so it is safe to convert!

But first, I'm going to confirm that 'shotgun' and 'no_huddle' are binary.

In [17]:
print(pass_run['shotgun'].unique())
print(pass_run['no_huddle'].unique())

[0. 1.]
[0. 1.]


Yup, they're clean. Now we can update the data types of the features.

In [25]:
pass_run['down'] = pass_run['down'].astype(int)
pass_run['ydstogo'] = pass_run['ydstogo'].astype(int)
pass_run['yardline_100'] = pass_run['yardline_100'].astype(int)
pass_run['score_differential'] = pass_run['score_differential'].astype(int)
pass_run['qtr'] = pass_run['qtr'].astype(int)
pass_run['game_seconds_remaining'] = pass_run['game_seconds_remaining'].astype(int)
pass_run['half_seconds_remaining'] = pass_run['half_seconds_remaining'].astype(int)
pass_run['quarter_seconds_remaining'] = pass_run['quarter_seconds_remaining'].astype(int)
pass_run['shotgun'] = pass_run['shotgun'].astype(int)
pass_run['no_huddle'] = pass_run['no_huddle'].astype(int)

In [ ]:
# let's verify
print(pass_run[initial_features].dtypes)

down                           int64
ydstogo                        int64
yardline_100                   int64
goal_to_go                     int64
score_differential             int64
qtr                            int64
game_seconds_remaining         int64
half_seconds_remaining         int64
game_half                     object
quarter_seconds_remaining      int64
home_timeouts_remaining      float64
away_timeouts_remaining      float64
shotgun                        int64
no_huddle                      int64
posteam                       object
dtype: object


Great, everything is looking clean! Now we can move onto the Feature Engineering.

To recap the Data Cleaning section, we:
- Filtered to run/pass plays only (34,628 rows)
- Identified and dropped 125 two-point conversion attempts (structural nulls in down)
- Confirmed shotgun and no_huddle are clean binary
- Converted all numeric columns from float64 to int64

### Part 3: Feature Engineering

In [27]:
import numpy as np

pass_run['posteam_timeouts_remaining'] = np.where(
    pass_run['posteam_type'] == 'home',
    pass_run['home_timeouts_remaining'],
    pass_run['away_timeouts_remaining']
)

pass_run['defteam_timeouts_remaining'] = np.where(
    pass_run['posteam_type'] == 'home',
    pass_run['away_timeouts_remaining'],
    pass_run['home_timeouts_remaining']
)

In [29]:
# verifying
print(pass_run[
    (pass_run['home_timeouts_remaining'] != 3) | 
    (pass_run['away_timeouts_remaining'] != 3)
][['posteam_type', 'home_timeouts_remaining', 'away_timeouts_remaining',
   'posteam_timeouts_remaining', 'defteam_timeouts_remaining']].head(20))

    posteam_type  home_timeouts_remaining  away_timeouts_remaining  \
151         home                      2.0                      3.0   
152         home                      2.0                      3.0   
153         home                      2.0                      3.0   
154         home                      2.0                      3.0   
155         home                      2.0                      3.0   
156         home                      2.0                      3.0   
157         home                      2.0                      3.0   
158         home                      2.0                      3.0   
162         away                      2.0                      3.0   
164         away                      1.0                      3.0   
166         away                      0.0                      3.0   
168         home                      0.0                      3.0   
169         home                      0.0                      3.0   
170         home    

Looking good! The number of home timouts and away timeouts stay consistent, regardless of who has the ball.

Now I have to convert them to ints and drop the old features ('posteam_timeouts_remaining' and defteam_timeouts_remaining).

### Part 4: Modeling

In [ ]:
%pip install pycaret
from pycaret.classification import *

model_setup = setup(data = df, target = "Churn")
best_model = compare_models()

Data cleaning -> feature engineering -> ML -> visualizations -> interactive feature (could be separate file in .py) -> strong writeup (look at the video) -> include business sense

data pipeline for constant updating, SQL integraeion for data cleaning or pipeline or something

double check the steps

double check if you need game time and half time becuase you already have wuarter and wuarter time

look at the classifiers (posteam vs home team) and make sure that they are all lined up (feature enigneering)